### PR TITLE
Add default column for missing features

### PR DESCRIPTION
**Pull request by rakeshkashyap123**
If the add.default.col.for.missing.data flag is turned on, this commit will add a null value to the features if data is missing for it. If the feature has a default value configured, the default value column will be added.

Also, we need to ensure that setting the flag is done through a single spark session. One of the SWA test was setting this flag but using a different spark session which caused test failures.

### CODE REVIEW

File: feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala [hunk 1]
```diff
@@ -386,6 +386,146 @@ class AnchoredFeaturesIntegTest extends FeathrIntegTest {
...
+
+    df.data.show()
+    val featureList = df.data.collect().sortBy(row => if (row.get(0) != null) row.getAs[String]("a_id") else "null")
+    assertEquals(featureList(0).getAs[Row]("aEmbedding"),
+      Row(mutable.WrappedArray.make(Array("")), mutable.WrappedArray.make(Array(2.0f))))
+    assertEquals(featureList(0).getAs[Row]("featureWithNull3"), "null")
+    assertEquals(featureList(0).getAs[Row]("featureWithNull5"), mutable.Map("" -> 1.0f))
+    assertEquals(featureList(0).getAs[Row]("featureWithNull"),-1.0f)
+    assertEquals(featureList(0).getAs[Row]("featureWithNull4"),Map())
+    assertEquals(featureList(0).getAs[Row]("featureWithNull2"),1.0f)
```
Comment (anirudhagar13) [inline feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala:521]: What happens when we try to derive a feature from a feature that has default values such as a  null or empty Map? Are those already handled?

File: feathr-impl/src/main/scala/com/linkedin/feathr/offline/join/workflow/AnchoredFeatureJoinStep.scala [hunk 2]
```diff
@@ -39,8 +40,62 @@ private[offline] class AnchoredFeatureJoinStep(
...
+    val defaults = missingFeatures.flatMap(s => s._2.featureAnchor.defaults)
+
+    // Create a map of feature to their feature type if configured.
+    val featureTypes = missingFeatures
+      .map(x => Some(x._2.featureAnchor.featureTypeConfigs))
+      .foldLeft(Map.empty[String, FeatureTypeConfig])((a, b) => a ++ b.getOrElse(Map.empty[String, FeatureTypeConfig]))
+
+    // We try to guess the column data type from the configured feature type. If feature type is not present, we will default to
+    // default feathr behavior of returning a map column of string to float.
+    val updatedDf = missingFeatures.keys.foldLeft(dataframe) { (observationDF, featureName) =>
```
[the remark this answers is not in the document]
Comment (rakeshkashyap123) [reply feathr-impl/src/main/scala/com/linkedin/feathr/offline/join/workflow/AnchoredFeatureJoinStep.scala:65]: Done.

File: feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala [hunk 3]
```diff
@@ -386,6 +386,146 @@ class AnchoredFeaturesIntegTest extends FeathrIntegTest {
...
+          |   featureList: ["featureWithNull", "derived_featureWithNull", "featureWithNull2", "featureWithNull3", "featureWithNull4",
+          |    "featureWithNull5", "derived_featureWithNull2", "featureWithNull6",
+          |    "aEmbedding", "memberEmbeddingAutoTZ"]
+          | }
+      """.stripMargin,
+      featureDefAsString =
+        """
+          | sources: {
+          |  swaSource: {
+          |    location: { path: "generaion/daily" }
```
[the remark this answers is not in the document]
Comment (rakeshkashyap123) [reply feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala:419]: It is misspelt on purpose.

File: feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala (hunk 1 above)
Comment (rakeshkashyap123) [reply feathr-impl/src/test/scala/com/linkedin/feathr/offline/AnchoredFeaturesIntegTest.scala:526]: Added a test case for that.

**Review by anirudhagar13 — APPROVED** [review]
Thanks for the fix.

**Pull request merged**